### ISSUE TITLE
Does not work under Windows

### ISSUE THREAD

**Issue by riccardobonatiaesys** [globalsign/est#32]
Hello,

i succesfully compile estclient under Ubuntu (amd64) and Raspbian (armv6l) but seems that i can't compile estclient under Windows; this is the output i get:

C:\Windows\System32>go version
go version go1.22.0 windows/amd64

C:\Windows\System32>go install github.com/globalsign/est/cmd/estclient@latest
github.com/ThalesIgnite/crypto11
C:\Users\xxx.yyy\go\pkg\mod\github.com\!thales!ignite\crypto11@v1.2.1\crypto11.go:120:16: undefined: pkcs11.ObjectHandle
C:\Users\xxx.yyy\go\pkg\mod\github.com\!thales!ignite\crypto11@v1.2.1\crypto11.go:139:22: undefined: pkcs11.ObjectHandle
C:\Users\xxx.yyy\go\pkg\mod\github.com\!thales!ignite\crypto11@v1.2.1\crypto11.go:167:14: undefined: pkcs11.Ctx
C:\Users\xxx.yyy\go\pkg\mod\github.com\!thales!ignite\crypto11@v1.2.1\crypto11.go:170:16: undefined: pkcs11.TokenInfo
C:\Users\xxx.yyy\go\pkg\mod\github.com\!thales!ignite\crypto11@v1.2.1\crypto11.go:176:27: undefined: pkcs11.SessionHandle
C:\Users\xxx.yyy\go\pkg\mod\github.com\!thales!ignite\crypto11@v1.2.1\aead.go:56:64: undefined: pkcs11.Mechanism
C:\Users\xxx.yyy\go\pkg\mod\github.com\!thales!ignite\crypto11@v1.2.1\aead.go:56:83: undefined: pkcs11.GCMParams
C:\Users\xxx.yyy\go\pkg\mod\github.com\!thales!ignite\crypto11@v1.2.1\attributes.go:15:25: undefined: pkcs11.Attribute
C:\Users\xxx.yyy\go\pkg\mod\github.com\!thales!ignite\crypto11@v1.2.1\sessions.go:34:17: undefined: pkcs11.Ctx
C:\Users\xxx.yyy\go\pkg\mod\github.com\!thales!ignite\crypto11@v1.2.1\sessions.go:35:16: undefined: pkcs11.SessionHandle
C:\Users\xxx.yyy\go\pkg\mod\github.com\!thales!ignite\crypto11@v1.2.1\crypto11.go:139:22: too many errors

i am not very skilled with go.

How can it be?

Thank you very much

**Comment by riccardobonatiaesys** [comment]
Hi @toddgaunt-gs , thanks for the suggestion; I tried compiling with WSL under Windows and it gave me the same strange errors that appear when compiling directly under Windows.

**Comment by toddgaunt-gs** [comment]
@riccardobonatiaesys sorry I'm not sure I'll be able to help very much as I don't have a Windows machine to rule that out as an issue. It looks like one of the dependencies of this project (github.com/thalesignite/pkcs11) isn't able to compile correctly on your setup. I would suggest trying this out in a clean environment with a fresh installation of Go. I have tried go1.22.0 locally to make sure it wasn't an issue with the version of Go you were using and still didn't encounter the same error:
```
todd.gaunt@fedora[~/go/pkg]
(2)> go1.22.0 install github.com/globalsign/est/cmd/estclient@latest                                                                                                                                                                                    [130]
go: downloading github.com/globalsign/est v1.0.6
go: downloading github.com/globalsign/tpmkeys v1.0.3
go: downloading github.com/google/go-tpm v0.3.2
go: downloading go.mozilla.org/pkcs7 v0.0.0-20200128120323-432b2356ecb1
go: downloading golang.org/x/crypto v0.0.0-20200602180216-279210d13fed
go: downloading github.com/ThalesIgnite/crypto11 v1.2.1
go: downloading github.com/globalsign/pemfile v1.0.0
go: downloading github.com/go-chi/chi v4.1.2+incompatible
go: downloading golang.org/x/time v0.0.0-20210220033141-f8bda1e9f3ba
go: downloading github.com/thales-e-security/pool v0.0.1
go: downloading github.com/pkg/errors v0.8.1
go: downloading github.com/miekg/pkcs11 v1.0.3-0.20190429190417-a667d056470f
go: downloading golang.org/x/sys v0.0.0-20210220050731-9a76102bfb43
```

**Comment by riccardobonatiaesys** [comment]
@toddgaunt-gs I tried to do a clean installation without success; now I will try with an earlier version of Windows and i will update you.

**Comment by 100110010111** [comment]
Just wanted to say I had the same problem on Windows. Would love to get this working.

**Comment by toddgaunt-gs** [comment]
Same! I don't have any Windows machines to test with so contribution would be welcome here.

**Comment by riccardobonatiaesys** [comment]
Yes guys, sorry for the late reply...I ran tests on Windows 11 and Windows 10, both without success. I will run tests with Windows 8.1 and Windows 7 shortly. I will keep you updated.

**Comment by toddgaunt-gs** [comment]
No need to apologize! Thank you for trying it out

**Comment by 100110010111** [comment]
What is the value of testing on OS that no one uses?

**Comment by riccardobonatiaesys** [comment]
> What is the value of testing on OS that no one uses?

I would just like to compile the executable for later use on windows 10/11.

**Comment by riccardobonatiaesys** [comment]
Back after a while without good news: also with Windows 8.1 the error persist. What can be done to resolve the situation?

**Comment by andreapiccaluga** [comment]
Hi guys, just a heads up about this issue. I'm an absolute newbie about Go, so I can't provide deeper insights, but I managed to compile the client for Windows on a Linux machine by installing a MinGW cross-compiler and then passing it as CC environment variable together with CGO_ENABLED=1 which apparently is a requirement for the library that gives all those errors. Maybe if you are on Windows already, you just need CGO_ENABLED=1, I don't know, feel free to experiment. Here's the command line that worked for me:

`CC=/usr/bin/x86_64-w64-mingw32-gcc CGO_ENABLED=1 GOOS=windows GOARCH=amd64 go build -ldflags='-s -w -extldflags "-static"' -o estclient.exe ./cmd/estclient`

**Comment by riccardobonatiaesys** [comment]
It works, thank you very much! To compile directly under Windows, simply change the environment variables of go using the following command: “go env -w CGO_ENABLED=1”.
